src/crypto/verification/Base.js

```diff
@@ -151,12 +151,16 @@ export default class VerificationBase extends EventEmitter {
             this._rejectEvent = undefined;
             this._resetTimer();
             this._resolveEvent(e);
+        } else if (e.getType() === "m.key.verification.cancel") {
+            const reject = this._reject;
```

Comment:
I think that if you just do a `this.cancel(e)`, then it should do the right thing, because the `cancel` function says that it handles cancellations from the other user (if it gets passed a `MatrixEvent`).  Though we probably also want to call `this._rejectEvent` if it's defined, so it may be better to just stick this in with the `else` clause, and assign to `exception` based on whether `e.getType() === "m.key.verification.cancel"`.

Though it looks like the `else` clause isn't doing the right thing right now, because it really should be causing the client to send an `m.key.verification.cancel` with `code` set to `m.unexpected_message`.  So, I think if you do something like

```javascript
const exception = e.getType() === "m.key.verification.cancel" ? e : newUnexpectedMessageError();
```

in the `else` clause (you'll need to import `newUnexpectedMessageError` from `Error.js`), then it should fix that and the double-cancel issue.

Reply:
Hmm, I tried calling `_rejectEvent`, but that made it send an `m.key.verification.cancel` event like before. Will have another look